docs/dev/deployment.md

```diff
@@ -1,18 +1,52 @@
-### Deployment
+### Release & Deployment
 
-**Only changes to `container/` and `Dockerfile`** trigger a build
+Note: **Only changes to `container/` and `Dockerfile`** will trigger a release.
 
-- To deploy to test network, just push to `main`.
-- To deploy to main network, create a tag and push it, example:
-  ```bash
-  git checkout main
-  git pull
-  git tag $(date +%s)
-  git push --follow-tags
-  ```
-
-In development, to avoid an automatic CI/CD deployment to the test network when any change is made to the `container/` directory, include `[skip ci]` in the `git commit` message. Like:
+In development, to avoid an automatic CI/CD release and subsequent deployment to the `test network` when a change is made to the `container/**` directory, include `[skip ci]` in the `git commit` message. Like so:
 
 ```bash
 git commit -m "my commit message [skip ci]"
 ```
+
+Staged deployment changes are meant to be rolled as depicted below:
+
+`canary subnetwork` -> `test network` -> `main network`
```

Comment:
Test network should go first, then canary which is a subset of main

Reply:
Sadly, it can't work like that or we'll have to run a long-lived `test` branch to have isolation. Any changes merged to the `main` branch will end up in both `main` and `test` networks.

Reply:
why can't we keep `main` branch as test, `canary` branch as canary and once tagged in `main` deployed to main?

Reply:
Because once a change ends up on main, it's going to share the same history as any upcoming changes. If a hotfix needs to be deployed the main network and we already have canary changes on main, then there is no way to avoid deploying those canary changes to main, as they linearly precede it.

Reply:
I agree, canary (mainnet) -> testnet -> mainnet is awkward. It should be testnet -> canary -> mainnet.

Reply:
It's not often we hotfix, and we can easily play with git if required. Let's no scramble it

Reply:
I agree it's awkward as well. 

Hotfixes will come and that approach stops us from fixing main in an unlikely but realistic scenario.
If we are to simply merge to main, then we defeat the purpose of a canary branch. If isolation does not matter, then this could be a simple canary tag.

Either way, isolation matters.
If this flow seems counter-intuitive then the future-proof way is to have a long-running test branch. I don't see any other reliable option 🤔

Reply:
But isn't the current GH action workflow describe the way we want it?

1. push to main to deploy to test
2. merge main into canary to deploy to canary nodes
3. tag main to deploy to l1s.

This works as expected

Reply:
So the flow should 100% be testnet -> canary -> mainnet. That doesn't mean we have to push canary changes to `main` branch in order to update testnet. That will have to change. Relying on git surgery to push hotfixes is a recipe for disaster imo. Our branches should be clean and easily hot fixable.

> Sadly, it can't work like that unless we have a long-lived test branch to have isolation (this means one branch per network and more maintenance overhead).

I'm leaning towards this now, and tbh the added maintenance doesn't sound so bad. We're already adding 2 PRs to our workflow. Here's my proposal, instead of a "long lived test branch", we make it the default branch.

Proposal
1. `dev`
  * This is now the default branch, do all your work here, and make PR branches from this one.
  * Pushes will update testnet. We are free to mess up and crash testnet, it's fine.
  * Make PRs to canary 
3. `canary`
  * PR merge & push will update canary
  * Make PRs to main
4. `main`
  * PR merge & push will update mainnet
  * Direct commits are fine for trivial changes? Documentation, grammar fix, etc.

Reply:
> merge main into canary to deploy to canary nodes

This is a bit backwards. Experimental changes should be isolated from the deploy branch.

> But isn't the current GH action workflow describe the way we want it?

Eh, we "want" it because it's convenient, not because its the best workflow. Committing arbitrary code into the same branch that you deploy from is not safe. I think that's our biggest issue atm.

Reply:
> So the flow should 100% be testnet -> canary -> mainnet. Any other order doesn't make sense. That doesn't mean we have to push canary changes to `main` branch in order to update testnet. That will have to change. Relying on git surgery to push hotfixes is a recipe for disaster imo. Our branches should be clean and easily hot fixable.

I agree `testnet -> canary -> mainnet` makes the most sense. In fact, that was what was aiming at when I started looking into this. Then I realized it wouldn't be possible and safe with a simple `canary` branch.
Absolutely, ideally this should all be automated and rely on as little human decisions as possible, e.g. avoiding git surgery.

> 
> > Sadly, it can't work like that unless we have a long-lived test branch to have isolation (this means one branch per network and more maintenance overhead).
> 
> I'm leaning towards this now, and tbh the added maintenance doesn't sound so bad. We're already adding 2 PRs to our workflow. 1 branch per env is an industry standard, no? Also, committing directly to `master/main` is an industry sin, ya?

Not necessarily. It depends on one's software delivery philosophy. Some years ago, it was standard to have one branch per environment. As DevOps practices, such as Continuous Delivery (CD) emerged, it became common to use a single mainline branch and release as often as possible with tagging and/or feature flags. The rationale is that, if you iterate on fixes and ship all the time, changesets will be as small as possible and less likely to cause complex problems. Thus, the impact of each change will be as self-contained and easy to reason about as possible. 
To accomplish this, teams typically performs Continuous Integration (CI) to make sure the software does what it's supposed to do and that there are no regressions (i.e. reappearing faults).

From my personal experience, minimizing the number of layers one has to go through to get a change to its users, should be a team goal. It's much more satisfying to get something out there (with test coverage ofc) than to delay it through processes which will often involve human validation.

I like your solution @guanzo. I just came back from my dance class (😂 ) and while exercising (🙌) I started leaning towards the same approach (i.e. having a dev/development branch), but with a tweak to cut on overhead. 

I worry about having multiple PRs between branches (i.e. `dev` -> `canary` -> `main`). Having worked with a similar kind of setup before, I know it can be demotivating and even frustrating as entropy irrevocably makes its way. Every time a PR is delayed, things get just a little bit more complex. Over time, branches differ for subtle reasons, that trickles, and then merge chaos ensues.
Sometimes, things work fine on one branch (e.g. development) and then break when integrated into a another branch (e.g. production) - this can happen due to configuration changes for instance. This means having to CI between all these branch PRs to be safe. These delays accumulate and suddenly things are less smooth.
 
> Here's my proposal, instead of a "long lived test branch", we make it the default branch.
Sounds good.
 
>     1. `dev`
> 
>     * This is now the default branch, do all your work here, and make PR branches from this one. 
>     * Pushes will update testnet. We are free to mess up and crash testnet, it's fine.
Completely agree so far 👍 .
Just a small note that Station relies on our `test network`.

>     * Make PRs to canary
I would like to avoid this step. More below.

>     * TBD: Trivial changes should just be merged to main? Skip PR.
This can bite us, but again we have been doing a good job here, so I wouldn't worry.
 
>     2. `main`
> 
>     * PR merge & push will update mainnet
sounds great
>     * TBD: Direct commits are fine for trivial changes? Documentation, grammar fix, etc.
Same as for the other TBD above.

Notice I removed the `canary` branch. What if tags automatically released changes (like the way we're doing now with `main`) to get deployed to the `canary subnet`? 

We don't really need isolation between our default/development branch and our production-like testing environment. We're testing stuff. `testnet`/development changes must land on `canarynet` at some point. However, these changes aren't ready for `mainnet`. There will be a PR when they are.
What if a change isn't worth testing on `canarynet`? We don't tag it and open a PR to merge to `main`.
What If `mainnet` lags behind a few Rhea iterations? That's completely fine, as some of these changes can be regarded as experimental. We'll release when things have been validated and are deemed stable. TBD what that means in practice.
One less branch. Better DX. No extra maintenance cost to pay. No extra CI server runs. Less procedural delay.

We do need to ensure `mainnet` is stable, especially in the face of risky changes. To do so, we isolate releases and deployments to this environment with its own branch (which shouldn't require any tag for the sake of simplicity).

What do you all think?

p.s. my only concern with the concept of staged deployments (in whatever shape) is the potential integration chaos that can result from running radically different L1 software pieces on the same network. Ideally, this should never happen, but entropy likes to have its way.

Reply:
> As DevOps practices, such as Continuous Delivery (CD) emerged, it became common to use a single mainline branch and release as often as possible with tagging and/or feature flags

This sounds nice. I think my only issue is that sometimes feature flags are a PITA to implement. But when they're not, could we mix branching + feature flags? We sort of do already, e.g. using Lassie when a request has `bifrost-gateway` user agent. I suppose flags are a natural use case in a lot of situations, we don't really need to think about it as hard as we do about our branching + deployment models.

> Having worked with a similar kind of setup before, I know it can be demotivating and even frustrating as entropy irrevocably makes its way.

Ya. Easy to say, hard to do in practice. 

> Notice I removed the canary branch. What if tags automatically released changes (like the way we're doing now with main) to get deployed to the canary subnet?

I like it. So I understand correctly:

* `dev` push = deploy testnet
* `dev` tag = deploy canary
* `main` push (no tag needed?) = deploy mainnet

Is that what you're proposing?

> We don't really need isolation between our default/development branch and our production-like testing environment. We're testing stuff. 

The caveat being that canary net is still receiving production traffic, so if we do break something, it has to have minimal impact. 

> What if a change isn't worth testing on canarynet? We don't tag it and open a PR to merge to main.

agree

> What If mainnet lags behind a few Rhea iterations? That's completely fine, as some of these changes can be regarded as experimental. We'll release when things have been validated and are deemed stable. TBD what that means in practice.

agree

> One less branch. Better DX. No extra maintenance cost to pay. No extra CI server runs. Less procedural delay.

hard agree!

> my only concern with the concept of staged deployments (in whatever shape) is the potential integration chaos that can result from running radically different L1 software pieces on the same network. 

Ya. We'll just need to be mindful of this. Keep the interfaces stable.

Reply:
> > As DevOps practices, such as Continuous Delivery (CD) emerged, it became common to use a single mainline branch and release as often as possible with tagging and/or feature flags
> 
> This sounds nice. I think my only issue is that sometimes feature flags are a PITA to implement. But when they're not, could we mix branching + feature flags? We sort of do already, e.g. using Lassie when a request has `bifrost-gateway` user agent. I suppose flags are a natural use case in a lot of situations, we don't really need to think about it as hard as we do about our branching + deployment models.
> 
Yeah I think feature flags are a more natural approach for delivering changes. They can be hard to implement, but IMO they put the burden of isolating the change in the development process itself and not on the branching strategy (which typically only adds extra complexity).

> > From my personal experience, minimizing the number of layers one has to go through to get a change to its users, should be a team goal. It's much more satisfying to get something out there (with test coverage ofc) than to delay it through processes which will often involve human validation.
> 
> Agree. That should be our goal.
> 
> > Having worked with a similar kind of setup before, I know it can be demotivating and even frustrating as entropy irrevocably makes its way.
> 
> Ya. Easy to say, hard to do in practice.
> 
> > This can bite us, but again we have been doing a good job here, so I wouldn't worry.
> 
> I don't want to get bitten :(. I'd be ok with a PR for all code changes. But for example, something like a trivial README/docs/comment change shouldn't be a PR imo, that seems like a time waster. At the same time, allowing exceptions here and there can cause trouble. I'm unsure about the path forward here.
> 
Yeah this part is tricky, but I wouldn't worry too much rn.

> > Notice I removed the canary branch. What if tags automatically released changes (like the way we're doing now with main) to get deployed to the canary subnet?
> 
> I like it. So I understand correctly:
> 
>     * `dev` push = deploy testnet
> 
>     * `dev` tag = deploy canary
> 
>     * `main` tag = deploy mainnet
> 
> 
> Is that what you're proposing?

Yap that's precisely it :)

> 
> > We don't really need isolation between our default/development branch and our production-like testing environment. We're testing stuff.
> 
> The caveat being that canary net is still receiving production traffic, so if we do break something, it has to have minimal impact.
> 
> > What if a change isn't worth testing on canarynet? We don't tag it and open a PR to merge to main.
> 
> agree
> 
> > What If mainnet lags behind a few Rhea iterations? That's completely fine, as some of these changes can be regarded as experimental. We'll release when things have been validated and are deemed stable. TBD what that means in practice.
> 
> agree
> 
> > One less branch. Better DX. No extra maintenance cost to pay. No extra CI server runs. Less procedural delay.
> 
> hard agree!
> 
> > my only concern with the concept of staged deployments (in whatever shape) is the potential integration chaos that can result from running radically different L1 software pieces on the same network.
> 
> Ya. We'll just need to be mindful of this. Keep the interfaces stable.

Reply:
Not trying to derail your current work, but how would the dev process look like if we went all in on feature flags? Is there anything more besides using tags+feature flags to deploy canary/prod

Reply:
I think it would be something like:

1. Our work branch is called `main`;
  - all changes (including doc fixes) go here;
  - all commits are automatically deployed to the test network.
  
2. Whenever a staged deployment is necessary:
    - we open a PR adding a feature flag to isolate it from the remaining codebase, disabling it by default (knowing it will end up on `main` next time there's a release);
    - we push a `canary-*` tag;
    - this will generate a canary release which will be picked up by the canary subnetwork;
    - an ansible run is required to enable the features to be tested on the canary subnetwork (we can probably fix this by working on [this](https://github.com/filecoin-saturn/L1-node/issues/208);
    - we wait 24h (left to the judgement of the people validating the changes) to see if anything breaks or is picked up by monitoring;
    - if
        - things break, changes are isolated from the main network, as they are disabled by default with a feature flag. they can be iterated upon until they are deemed stable;
        - all is fine, we can move forward and release to the main network.

3. When ready to release to the main network
  - we tag the main branch with a non-canary tag, i.e. not `canary-*`.
  
The only other thing that comes to mind was somehow thinking about blue-green deployments as a tool to achieve this, but looking [at this](https://www.optimizely.com/insights/blog/feature-flags-blue-green-deployments/), feature flags seem a better candidate.